src/symbol/collision_index.js

```diff
@@ -0,0 +1,353 @@
+
+        const query = [];
+        let minX = Infinity;
+        let minY = Infinity;
+        let maxX = -Infinity;
+        let maxY = -Infinity;
+        for (let i = 0; i < queryGeometry.length; i++) {
+            const ring = queryGeometry[i];
+            for (let k = 0; k < ring.length; k++) {
+                const p = this.projectPoint(posMatrix, ring[k].x, ring[k].y);
+                minX = Math.min(minX, p.x);
+                minY = Math.min(minY, p.y);
+                maxX = Math.max(maxX, p.x);
+                maxY = Math.max(maxY, p.y);
+                query.push(p);
+            }
+        }
+
+        const tileID = tileCoord.id;
+
+        const thisTileFeatures = [];
+        const features = this.grid.query(minX, minY, maxX, maxY);
+        for (let i = 0; i < features.length; i++) {
+            if (features[i].tileID === tileID) {
+                thisTileFeatures.push(features[i].boxIndex);
+            }
+        }
+        const ignoredFeatures = this.ignoredGrid.query(minX, minY, maxX, maxY);
+        for (let i = 0; i < ignoredFeatures.length; i++) {
+            if (ignoredFeatures[i].tileID === tileID) {
+                thisTileFeatures.push(ignoredFeatures[i].boxIndex);
+            }
+        }
+
+        for (let i = 0; i < thisTileFeatures.length; i++) {
+            const blocking = collisionBoxArray.get(thisTileFeatures[i]);
+            const sourceLayer = blocking.sourceLayerIndex;
+            const featureIndex = blocking.featureIndex;
+
+            // Skip already seen features.
+            if (sourceLayerFeatures[sourceLayer] === undefined) {
+                sourceLayerFeatures[sourceLayer] = {};
+            }
+            if (sourceLayerFeatures[sourceLayer][featureIndex]) continue;
+
+
+            // Check if query intersects with the feature box
+            // "Collision Circles" for line labels are treated as boxes here
+            // Since there's no actual collision taking place, the circle vs. square
+            // distinction doesn't matter as much, and box geometry is easier
+            // to work with.
+            const projectedPoint = this.projectAndGetPerspectiveRatio(posMatrix, blocking.anchorPointX, blocking.anchorPointY);
+            const tileToViewport = textPixelRatio * projectedPoint.perspectiveRatio;
+            const x1 = blocking.x1 / tileToViewport + projectedPoint.point.x;
+            const y1 = blocking.y1 / tileToViewport + projectedPoint.point.y;
+            const x2 = blocking.x2 / tileToViewport + projectedPoint.point.x;
+            const y2 = blocking.y2 / tileToViewport + projectedPoint.point.y;
+            const bbox = [
+                new Point(x1, y1),
+                new Point(x2, y1),
+                new Point(x2, y2),
+                new Point(x1, y2)
+            ];
+            if (!intersectionTests.polygonIntersectsPolygon(query, bbox)) continue;
```

Comment:
Is this the only place where we use the `x1, y1, x2, y2` (box) data for the along-line-labels/collision-circle case?  If we added circle-aware handling here, could we then split `collisionBoxArray` into two separate arrays, one for boxes and one for (sets of) circles?

Reply:
Yes, we could, although I don't think having the extra bit of unused data in the `collisionBoxArray` is much of a performance concern. If we switch to foreground circle generation (#5474), then the `collisionBoxArray` as a whole will become almost entirely vestigial (it would really just hold indices for `queryRenderedSymbols`). At that point I think it would make sense to just remove it entirely.